src/datastructure/Mask2DCollection.cpp

```diff
@@ -0,0 +1,227 @@
+/**
+ * @copyright Copyright (c) 2017-2025 B-com http://www.b-com.com/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "datastructure/Mask2DCollection.h"
+#include "core/Log.h"
+#include <filesystem>
+#include <nlohmann/json.hpp>
+#include <xpcf/core/helpers.h>
+
+BOOST_CLASS_EXPORT_IMPLEMENT(SolAR::datastructure::Mask2DCollection);
+
+namespace xpcf = org::bcom::xpcf;
+using json = nlohmann::json;
+
+namespace SolAR {
+namespace datastructure {
+
+FrameworkReturnCode Mask2DCollection::addMask(SRef<Mask2D> mask, uint32_t& maskId)
+{
+    if (!mask) {
+        LOG_ERROR("Mask2DCollection::addMask - invalid input mask pointer.");
+        return FrameworkReturnCode::_ERROR_;
+    }
+    for (auto& [existingId, existingMask] : m_masks) {
```

Comment:
nit: Interesting check, but is it worth the cost ? With a lot of Frames, and the complexity of equals(), this might have an impact. Let's keep it and see if it's the case.
Also, can this never happen, for example when replaying a recorded sequence in a loop, we'll have the exact same images, likely leading to the exact same mask ? (but it's an edge case)